[Kyriel Mortel Abad] iP

s-kybound opened the pull request:
# Crusader

> “This life, this entire world, was a crucible. It was the crusade of their times, and they were the knights, the warriors.”
― Nancy Holder, Debbie Viguie 

A one-stop helper for all your needs!

Why Crusader? It's

- Made by Kyriel
- Elegant
- Has _class_ (several of them actually hahahaha 🤓)

To use it and join the crusader force,

1. Find the latest release [here](https://github.com/s-kybound/ip/releases)
2. Install it
3. Get Java
4. Open a CLI
5. Run the following command: 
```
java -jar crusader.jar
```
And enjoy!

## Features

- [x] Can track tasks
- [x] Can save and load tasks
- [ ] Reminders
- [x] Fast
- [ ] ~Astoundingly~ _Blazingly_ fast

## Example code

```java
/**
 * Activates the bot.
 */
public void run() {
    this.ui.showLogo();
    this.ui.greet();
    boolean hasEnded = false;
    while (!hasEnded) {
        try {
            String command = this.ui.promptInput();
            Command c = Parser.parse(command);
            c.execute(ui, taskList);
            hasEnded = c.isExit();
        } catch (CrusaderException e) {
            this.ui.say(e.getMessage());
        }
    }
    this.ui.farewell();
    this.storage.saveTasks(taskList.getTasks());
}

public static void main(String[] args) {
    new Crusader(SAVE_FILE, LOGO).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

s-kybound committed:
Personalise chatbot with new name

s-kybound committed:
Add horizontal divider

s-kybound committed:
Add greeter and goodbye to bot

s-kybound committed:
Add echo to bot

s-kybound committed:
Add list to bot

s-kybound committed:
Add task class to bot

s-kybound committed:
Add todo, marking and unmarking

s-kybound committed:
Add feedback to commands

s-kybound committed:
Initialize Task subclasses

s-kybound committed:
Integrate Task Subclasses into Crusader.java

s-kybound committed:
Updated testing

s-kybound committed:
Implement basic error handling

s-kybound committed:
Implement deletion of tasks

s-kybound committed:
Implement skeleton load and save functions

s-kybound committed:
Implement load and save

s-kybound committed:
Implement proper date and time representation

s-kybound committed:
Rename variables to conform to convention

s-kybound committed:
Merge branch 'branch-Level-7'

s-kybound committed:
Rename variables to follow convention

s-kybound committed:
Merge branch 'branch-Level-8'

s-kybound committed:
Implement TaskList

s-kybound committed:
Implement UI

s-kybound committed:
Implement prompting in UI

s-kybound committed:
Implement date utilites

s-kybound committed:
Implement storage utlilties

s-kybound committed:
Add constructor to Storage

s-kybound committed:
Implement Exceptions

s-kybound committed:
Integrate Date utilities into Tasks

s-kybound committed:
Implement alternative TaskList constructor

s-kybound committed:
Implement Command classes

s-kybound committed:
Implement Parser

s-kybound committed:
Clean up code

s-kybound committed:
Integrate OOP classes into Crusader

s-kybound committed:
Implement index range checks for array access

s-kybound committed:
Merge branch 'branch-A-MoreOOP'

s-kybound committed:
Reorganise code into packages

s-kybound committed:
Merge branch 'branch-A-Packages'

s-kybound committed:
Merge branch 'add-gradle-support' into A-Gradle

s-kybound committed:
Add Gradle support for Crusader

s-kybound committed:
Merge branch 'branch-A-Gradle'

s-kybound committed:
Implement JUnit testing

s-kybound committed:
Merge branch 'branch-A-JUnit'

s-kybound committed:
Implement find command

s-kybound committed:
Add preliminary Javadocs to all files

s-kybound committed:
Enforce CS2013T style convention

s-kybound committed:
Merge branch 'branch-A-JavaDoc'

s-kybound committed:
Merge branch 'branch-A-CodingStandard'

s-kybound committed:
Resolve merge conflict in Parser

s-kybound committed:
Rename variable in Crusader

s-kybound committed:
Merge branch 'branch-Level-9'

jinyang628 commented on src/main/java/crusader/DateUtils.java:
```diff
@@ -0,0 +1,46 @@
+     * @param date A date
+     * @return A string formatted for saving
+     */
+    public static String saveFormat(Date date) {
```
I love the way you use Egyptian braces throughout your program!

jinyang628 commented on src/main/java/crusader/Parser.java:
```diff
@@ -0,0 +1,194 @@
+package crusader;
+
+import crusader.command.AddTaskCommand;
+import crusader.command.ByeCommand;
```
I love the way you import the classes explicitly instead of using *

jinyang628 commented on src/main/java/crusader/Parser.java:
```diff
@@ -0,0 +1,194 @@
+        switch (prompt.contains(" ")
+                ? prompt.split(" ")[0]
+                : prompt) {
+        case "bye":
```
Good job adhering to the correct indentation levels for switch statements!

jinyang628 commented on src/main/java/crusader/Parser.java:
```diff
@@ -0,0 +1,194 @@
+     */
+    public static Command parse(String prompt) throws CrusaderException {
+        Command returnCommand;
+        switch (prompt.contains(" ")
```
I absolutely love the way you make line breaks to improve readability!

jinyang628 commented on src/main/java/crusader/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+    public String toString() {
+        String returnString = "";
+        for (int x = 0; x < tasks.size(); x++) {
+            returnString = String.format("%s%d. %s\n", returnString, x + 1, tasks.get(x).toString());
```
Love the way you adhere to the line limit and push extra characters to the next line

jinyang628 reviewed (commented):
Love the way you structure your program Kyriel! I learnt a lot from the way you code :)

s-kybound committed:
Implement checkstyle

s-kybound committed:
Resolve checkstyle issues

s-kybound committed:
Merge branch 'branch-A-Checkstyle'

s-kybound committed:
Implement first UI

s-kybound committed:
Clean up UI code

s-kybound committed:
Merge branch 'branch-Level-10'

s-kybound committed:
Implement assertions in Crusader

Many parts of Crusader code make implicit assumptions about
conditions that are required to let them execute properly.
This is unsafe as these conditions may no longer be true as
the project evolves.

Using assertions allows us to make such assumptions explicit, so that
contributors are better aware of the required conditions for our project
to behave as expected.

Let's allow for the use of assertions in the code, as well as begin to
use assertions throughout the project.

s-kybound committed:
Improve code quality by simplifying expresssions

Code in the Crusader project uses many non-trivial expressions.

This makes code less readable for contributors.

By explicitly elaborating on these expressions by using variables, it
will help make code more readable.

Let's simplify complex expressions by giving them names to code clearer.

s-kybound committed:
Merge pull request #3 from s-kybound/branch-A-CodeQuality

Improve code quality by simplifying expresssions

s-kybound committed:
Merge pull request #2 from s-kybound/branch-A-Assertions

Implement assertions in Crusader

s-kybound committed:
Implement sorting in Crusader

s-kybound committed:
Merge pull request #4 from s-kybound/branch-BCD-Extension

Implement sorting in Crusader

s-kybound committed:
Improve GUI

s-kybound committed:
Merge branch 'branch-A-BetterGui'

s-kybound committed:
Remove redundant code

s-kybound committed:
Add README

s-kybound committed:
Merge remote-tracking branch 'origin/branch-A-UserGuide'

s-kybound committed:
Add picture

s-kybound committed:
Merge pull request #5 from s-kybound/branch-A-UserGuide

Add picture
